### DISCUSSION TITLE
RFC Draft: Expose Javascript API for CLI commands

### DISCUSSION THREAD

**Discussion by AgentEnder** [withastro/roadmap#47] [Proposal]
## Background & Motivation
I want to be able to programmatically invoke the CLI commands. Currently, the only way to consume them is through the CLI, the different functions to load the config, build, start the dev server, and start the preview server are not exported by the package.

A Javascript API would be particularly useful for scenarios where we have custom scripts or tooling to invoke these operations. In these scenarios we might need to manipulate the config once is loaded and before invoking the functions to build or start the dev or preview servers.

## Proposed Solution
Export the functions build, dev, preview, reload and loadConfig.

## Alternatives considered
Create a temporary config on the fly before spawning a process to invoke the CLI and feeding the temporary config to it. This is quite cumbersome and not worthy when a simple Javascript API can be provided.

## Risks, downsides, and/or tradeoffs
The only risk I can think of is about committing to a somewhat stable API. But I'd say the CLI API should be pretty stable at this point, the implementation is changing but I guess the signature for each function is still quite similar. Even if that's not exactly the case, given Astro is in an early beta stage, changes are expected to happen, so, it should be fine.

Copied over from https://github.com/withastro/astro/issues/1483

**Comment by leosvelperez** [comment]
hey @matthewp!

Sorry for the ping. You seemed on board with this idea when it was first suggested https://github.com/withastro/astro/issues/1483#issuecomment-941144677. Any chance we can move this forward?

This is very much needed for external tooling. I maintain an Astro plugin for Nx (https://github.com/nxtensions/nxtensions/tree/main/packages/astro) and having a Javascript API would simplify things a lot and allow for better integration. Other functions might also be useful to expose, but I think it would be a good start if we expose the CLI commands and utilities to resolve and load the Astro config.

**Comment by nicholasray** [comment]
+1 Would love to see this!

**Comment by florian-lefebvre** [comment]
We have an experimental programmatic API: https://docs.astro.build/en/reference/programmatic-reference/
